File: ts-sdk/src/protocol/crypto.ts
/**
 * Cryptographic primitives for UAM protocol.
 *
 * Wraps libsodium-wrappers for Ed25519 signing, Curve25519 key exchange,
 * NaCl Box (authenticated encryption), and NaCl SealedBox (anonymous encryption).
 *
 * This module never hand-rolls crypto -- every operation delegates to libsodium.
 */

import sodium from "libsodium-wrappers";
import { createHash } from "node:crypto";

import {
  DecryptionError,
  EncryptionError,
  SignatureVerificationError,
} from "./errors.js";
import { b64Decode, b64Encode } from "./types.js";

// ---------------------------------------------------------------------------
// Sodium initialization
// ---------------------------------------------------------------------------

/**
 * Promise that resolves when libsodium is ready.
 * Callers must await this before first use of any crypto function.
 */
export const sodiumReady: Promise<void> = sodium.ready;

// ---------------------------------------------------------------------------
// Key type aliases
// ---------------------------------------------------------------------------

/** 64-byte Ed25519 secret key (seed + public). */
export type SigningKey = Uint8Array;

/** 32-byte Ed25519 public key. */
export type VerifyKey = Uint8Array;

/** 32-byte seed. */
export type Seed = Uint8Array;

// ---------------------------------------------------------------------------
// Key generation and serialization
// ---------------------------------------------------------------------------

export interface Keypair {
  signingKey: Uint8Array;
  verifyKey: Uint8Array;
  seed: Uint8Array;
}

/**
 * Generate an Ed25519 keypair.
 */
export function generateKeypair(): Keypair {
  const seed = sodium.randombytes_buf(32);
  const kp = sodium.crypto_sign_seed_keypair(seed);
  return {
    signingKey: kp.privateKey,
    verifyKey: kp.publicKey,
    seed,
  };
}

/**
 * Serialize a signing key to URL-safe base64 (32-byte seed).
 *
 * Matches Python: PyNaCl's SigningKey.encode() returns the 32-byte seed.
 */
export function serializeSigningKey(seed: Uint8Array): string {
  return b64Encode(seed);
}

/**
 * Restore a signing key from its base64-encoded seed.
 */
export function deserializeSigningKey(s: string): Keypair {
  const seed = b64Decode(s);
  const kp = sodium.crypto_sign_seed_keypair(seed);
  return {
    signingKey: kp.privateKey,
    verifyKey: kp.publicKey,
    seed,
  };
}

/**
 * Serialize a verify (public) key to URL-safe base64.
 */
export function serializeVerifyKey(key: Uint8Array): string {
  return b64Encode(key);
}

/**
 * Restore a verify key from its base64 encoding.
 */
export function deserializeVerifyKey(s: string): Uint8Array {
  return b64Decode(s);
}

/**
 * Return the SHA-256 hex digest of the verify key bytes.
 *
 * This 64-character string serves as the agent's identity fingerprint.
 */
export function publicKeyFingerprint(verifyKey: Uint8Array): string {
  return createHash("sha256").update(Buffer.from(verifyKey)).digest("hex");
}

// ---------------------------------------------------------------------------
// Canonical JSON
// ---------------------------------------------------------------------------

/**
 * Recursively sort all object keys and produce compact JSON.
 *
 * Matches Python: json.dumps(data, sort_keys=True, separators=(',',':'), ensure_ascii=True)
 */
function sortedStringify(value: unknown): string {
  if (value === null || value === undefined) {
    return "null";
  }
  if (typeof value === "string") {
    return JSON.stringify(value);
  }
  if (typeof value === "number" || typeof value === "boolean") {
    return String(value);
  }
  if (Array.isArray(value)) {
    const items = value.map((v) => sortedStringify(v));
    return "[" + items.join(",") + "]";
  }
  if (typeof value === "object") {
    const obj = value as Record<string, unknown>;
    const keys = Object.keys(obj).sort();
    const pairs = keys.map((k) => JSON.stringify(k) + ":" + sortedStringify(obj[k]));
    return "{" + pairs.join(",") + "}";
  }
  return String(value);
}

/**
 * Produce deterministic JSON bytes for signing.
 *
 * - Excludes the "signature" key.
 * - Excludes keys whose value is null or undefined.
 * - Sorts keys, uses compact separators, ensures ASCII encoding.
 */
export function canonicalize(data: Record<string, unknown>): Uint8Array {
  const filtered: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(data)) {
    if (k !== "signature" && v !== null && v !== undefined) {
      filtered[k] = v;
    }
  }
  const json = sortedStringify(filtered);
  return new TextEncoder().encode(json);
}

// ---------------------------------------------------------------------------
// Signing and verification
// ---------------------------------------------------------------------------

/**
 * Sign data with the Ed25519 signing key.
 *
 * @returns The 64-byte signature as a URL-safe base64 string.
 */
export function signMessage(data: Uint8Array, signingKey: Uint8Array): string {
  const signature = sodium.crypto_sign_detached(data, signingKey);
  return b64Encode(signature);
}

/**
 * Verify an Ed25519 signature.
 *
 * @throws {SignatureVerificationError} If the signature is invalid.
 */
export function verifySignature(
  data: Uint8Array,
  signatureB64: string,
  verifyKey: Uint8Array
): void {
  try {
    const sigBytes = b64Decode(signatureB64);
    const valid = sodium.crypto_sign_verify_detached(sigBytes, data, verifyKey);
    if (!valid) {
      throw new SignatureVerificationError("Signature verification failed");
    }
  } catch (err) {
    if (err instanceof SignatureVerificationError) {
      throw err;
    }
    throw new SignatureVerificationError(
      err instanceof Error ? err.message : String(err)
    );
  }
}

// ---------------------------------------------------------------------------
// Nonce generation
// ---------------------------------------------------------------------------

/**
 * Generate 24 cryptographically random bytes, returned as base64.
 */
export function generateNonce(): string {
  return b64Encode(sodium.randombytes_buf(24));
}

// ---------------------------------------------------------------------------
// NaCl Box encryption (authenticated, both parties known)
// ---------------------------------------------------------------------------

/**
 * Encrypt plaintext using NaCl Box (authenticated encryption).
 *
 * Both parties are known; the sender signs implicitly via key exchange.
 * Format: nonce (24 bytes) || ciphertext, matching PyNaCl Box.encrypt() output.
 *
 * @returns Base64-encoded ciphertext (nonce prepended).
 * @throws {EncryptionError} On any libsodium error.
 */
export function encryptPayload(
  plaintext: Uint8Array,
  senderSigningKey: Uint8Array,
  recipientVerifyKey: Uint8Array
): string {
  try {
    const senderCurve = sodium.crypto_sign_ed25519_sk_to_curve25519(senderSigningKey);
    const recipientCurve = sodium.crypto_sign_ed25519_pk_to_curve25519(recipientVerifyKey);
    const nonce = sodium.randombytes_buf(sodium.crypto_box_NONCEBYTES);
    const encrypted = sodium.crypto_box_easy(plaintext, nonce, recipientCurve, senderCurve);
    // Concatenate nonce || ciphertext (matching PyNaCl format)
    const combined = new Uint8Array(nonce.length + encrypted.length);
    combined.set(nonce, 0);
    combined.set(encrypted, nonce.length);
    return b64Encode(combined);
  } catch (err) {
    if (err instanceof EncryptionError) throw err;
    throw new EncryptionError(err instanceof Error ? err.message : String(err));
  }
}

/**
 * Decrypt NaCl Box ciphertext.
 *
 * @returns The original plaintext bytes.
 * @throws {DecryptionError} On any libsodium error (wrong keys, tampered data).
 */
export function decryptPayload(
  ciphertextB64: string,
  recipientSigningKey: Uint8Array,
  senderVerifyKey: Uint8Array
): Uint8Array {
  try {
    const combined = b64Decode(ciphertextB64);
    const nonce = combined.slice(0, sodium.crypto_box_NONCEBYTES);
    const ciphertext = combined.slice(sodium.crypto_box_NONCEBYTES);
    const recipientCurve = sodium.crypto_sign_ed25519_sk_to_curve25519(recipientSigningKey);
    const senderCurve = sodium.crypto_sign_ed25519_pk_to_curve25519(senderVerifyKey);
    return sodium.crypto_box_open_easy(ciphertext, nonce, senderCurve, recipientCurve);
  } catch (err) {
    if (err instanceof DecryptionError) throw err;
    throw new DecryptionError(err instanceof Error ? err.message : String(err));
  }
}

// ---------------------------------------------------------------------------
// NaCl SealedBox encryption (anonymous sender)
// ---------------------------------------------------------------------------

/**
 * Encrypt plaintext using NaCl SealedBox (anonymous sender).
 *
 * Only the recipient's public key is required; no sender authentication.
 *
 * @returns Base64-encoded ciphertext.
 */
export function encryptPayloadAnonymous(
  plaintext: Uint8Array,
  recipientVerifyKey: Uint8Array
): string {
  const recipientCurve = sodium.crypto_sign_ed25519_pk_to_curve25519(recipientVerifyKey);
  const ciphertext = sodium.crypto_box_seal(plaintext, recipientCurve);
  return b64Encode(ciphertext);
}

/**
 * Decrypt NaCl SealedBox ciphertext.
 *
 * @returns The original plaintext bytes.
 * @throws {DecryptionError} On any libsodium error.
 */
export function decryptPayloadAnonymous(
  ciphertextB64: string,
  recipientSigningKey: Uint8Array
): Uint8Array {
  try {
    const ciphertext = b64Decode(ciphertextB64);
    const recipientCurveSk = sodium.crypto_sign_ed25519_sk_to_curve25519(recipientSigningKey);
    // Extract the public key from the signing key (last 32 bytes of the 64-byte Ed25519 secret key)
    const recipientVerifyKey = recipientSigningKey.slice(32);
    const recipientCurvePk = sodium.crypto_sign_ed25519_pk_to_curve25519(recipientVerifyKey);
    return sodium.crypto_box_seal_open(ciphertext, recipientCurvePk, recipientCurveSk);
  } catch (err) {
    if (err instanceof DecryptionError) throw err;
    throw new DecryptionError(err instanceof Error ? err.message : String(err));
  }
}

File: src/uam/protocol/crypto.py
"""Cryptographic primitives for UAM protocol.

Wraps PyNaCl (libsodium) for Ed25519 signing, Curve25519 key exchange,
NaCl Box (authenticated encryption), and NaCl SealedBox (anonymous encryption).

This module never hand-rolls crypto -- every operation delegates to PyNaCl.
"""

from __future__ import annotations

import hashlib
import json

import nacl.exceptions
import nacl.utils
from nacl.public import Box, SealedBox
from nacl.signing import SigningKey, VerifyKey

from uam.protocol.errors import (
    DecryptionError,
    EncryptionError,
    SignatureVerificationError,
)
from uam.protocol.types import b64_decode, b64_encode


# ---------------------------------------------------------------------------
# Key generation and serialization
# ---------------------------------------------------------------------------

def generate_keypair() -> tuple[SigningKey, VerifyKey]:
    """Generate an Ed25519 keypair.

    Returns:
        A ``(signing_key, verify_key)`` tuple.
    """
    sk = SigningKey.generate()
    return sk, sk.verify_key


def serialize_signing_key(key: SigningKey) -> str:
    """Serialize a signing key to URL-safe base64 (32-byte seed)."""
    return b64_encode(key.encode())


def deserialize_signing_key(s: str) -> SigningKey:
    """Restore a signing key from its base64-encoded seed."""
    return SigningKey(b64_decode(s))


def serialize_verify_key(key: VerifyKey) -> str:
    """Serialize a verify (public) key to URL-safe base64."""
    return b64_encode(key.encode())


def deserialize_verify_key(s: str) -> VerifyKey:
    """Restore a verify key from its base64 encoding."""
    return VerifyKey(b64_decode(s))


def public_key_fingerprint(verify_key: VerifyKey) -> str:
    """Return the SHA-256 hex digest of the verify key bytes.

    This 64-character string serves as the agent's identity fingerprint.
    """
    return hashlib.sha256(verify_key.encode()).hexdigest()


# ---------------------------------------------------------------------------
# Canonical JSON
# ---------------------------------------------------------------------------

def canonicalize(data: dict) -> bytes:
    """Produce deterministic JSON bytes for signing.

    - Excludes the ``"signature"`` key.
    - Excludes keys whose value is ``None``.
    - Sorts keys, uses compact separators, ensures ASCII encoding.
    """
    filtered = {
        k: v for k, v in data.items()
        if k != "signature" and v is not None
    }
    return json.dumps(
        filtered, sort_keys=True, separators=(",", ":"), ensure_ascii=True
    ).encode("utf-8")


# ---------------------------------------------------------------------------
# Signing and verification
# ---------------------------------------------------------------------------

def sign_message(data: bytes, signing_key: SigningKey) -> str:
    """Sign *data* with the Ed25519 *signing_key*.

    Returns:
        The 64-byte signature as a URL-safe base64 string.
    """
    signed = signing_key.sign(data)
    return b64_encode(signed.signature)


def verify_signature(
    data: bytes, signature_b64: str, verify_key: VerifyKey
) -> None:
    """Verify an Ed25519 signature.

    Raises:
        SignatureVerificationError: If the signature is invalid.
    """
    try:
        sig_bytes = b64_decode(signature_b64)
        verify_key.verify(data, sig_bytes)
    except nacl.exceptions.BadSignatureError as exc:
        raise SignatureVerificationError(str(exc)) from exc


# ---------------------------------------------------------------------------
# Nonce generation
# ---------------------------------------------------------------------------

def generate_nonce() -> str:
    """Generate 24 cryptographically random bytes, returned as base64."""
    return b64_encode(nacl.utils.random(24))


# ---------------------------------------------------------------------------
# NaCl Box encryption (authenticated, both parties known)
# ---------------------------------------------------------------------------

def encrypt_payload(
    plaintext: bytes,
    sender_signing_key: SigningKey,
    recipient_verify_key: VerifyKey,
) -> str:
    """Encrypt *plaintext* using NaCl Box (authenticated encryption).

    Both parties are known; the sender signs implicitly via key exchange.

    Returns:
        Base64-encoded ciphertext.

    Raises:
        EncryptionError: On any libsodium error.
    """
    try:
        sender_private = sender_signing_key.to_curve25519_private_key()
        recipient_public = recipient_verify_key.to_curve25519_public_key()
        box = Box(sender_private, recipient_public)
        ciphertext = box.encrypt(plaintext)
        return b64_encode(ciphertext)
    except nacl.exceptions.CryptoError as exc:
        raise EncryptionError(str(exc)) from exc


def decrypt_payload(
    ciphertext_b64: str,
    recipient_signing_key: SigningKey,
    sender_verify_key: VerifyKey,
) -> bytes:
    """Decrypt NaCl Box ciphertext.

    Returns:
        The original plaintext bytes.

    Raises:
        DecryptionError: On any libsodium error (wrong keys, tampered data).
    """
    try:
        recipient_private = recipient_signing_key.to_curve25519_private_key()
        sender_public = sender_verify_key.to_curve25519_public_key()
        box = Box(recipient_private, sender_public)
        return box.decrypt(b64_decode(ciphertext_b64))
    except nacl.exceptions.CryptoError as exc:
        raise DecryptionError(str(exc)) from exc


# ---------------------------------------------------------------------------
# NaCl SealedBox encryption (anonymous sender)
#
# Used by create_envelope() for handshake.request messages where the sender
# may not have an established relationship with the recipient.  The envelope
# signature still authenticates the sender; SealedBox just means the
# encryption itself does not require the sender's private key.
# ---------------------------------------------------------------------------

def encrypt_payload_anonymous(
    plaintext: bytes, recipient_verify_key: VerifyKey
) -> str:
    """Encrypt *plaintext* using NaCl SealedBox (anonymous sender).

    Only the recipient's public key is required; no sender authentication.

    Returns:
        Base64-encoded ciphertext.
    """
    recipient_public = recipient_verify_key.to_curve25519_public_key()
    sealed = SealedBox(recipient_public)
    ciphertext = sealed.encrypt(plaintext)
    return b64_encode(ciphertext)


def decrypt_payload_anonymous(
    ciphertext_b64: str, recipient_signing_key: SigningKey
) -> bytes:
    """Decrypt NaCl SealedBox ciphertext.

    Returns:
        The original plaintext bytes.

    Raises:
        DecryptionError: On any libsodium error.
    """
    try:
        recipient_private = recipient_signing_key.to_curve25519_private_key()
        sealed = SealedBox(recipient_private)
        return sealed.decrypt(b64_decode(ciphertext_b64))
    except nacl.exceptions.CryptoError as exc:
        raise DecryptionError(str(exc)) from exc
